readme.md
<p align="center">
  <img src="/images/wled_logo_akemi.png">
  <a href="https://github.com/atuline/WLED/releases"><img src="https://img.shields.io/github/release/atuline/WLED.svg?style=flat-square"></a>
  <a href="https://raw.githubusercontent.com/Aircoookie/WLED/master/LICENSE"><img src="https://img.shields.io/github/license/Aircoookie/wled?color=blue&style=flat-square"></a>
  <a href="https://wled.discourse.group"><img src="https://img.shields.io/discourse/topics?colorB=blue&label=forum&server=https%3A%2F%2Fwled.discourse.group%2F&style=flat-square"></a>
  <a href="https://discord.gg/4CQRmfR"><img src="https://img.shields.io/discord/700041398778331156.svg?colorB=blue&label=discord&style=flat-square"></a>
  <a href="https://github.com/atuline/WLED/wiki"><img src="https://img.shields.io/badge/quick_start-wiki-blue.svg?style=flat-square"></a>
  <a href="https://github.com/Aircoookie/WLED-App"><img src="https://img.shields.io/badge/app-wled-blue.svg?style=flat-square"></a>
  <a href="https://gitpod.io/#https://github.com/atuline/WLED"><img src="https://img.shields.io/badge/Gitpod-ready--to--code-blue?style=flat-square&logo=gitpod"></a>

  </p>

# Stable Branch
This is the SR `master` branch - the source code for our latest release version [SR WLED v0.13.2](https://github.com/atuline/WLED/releases/tag/v0.13.2).

- *This* branch (`master`) can be a stable baseline for your own project. Use it.
- *Pull Requests* should be created against our [`dev`elopment branch](https://github.com/atuline/WLED/tree/dev).



# Sound Reactive WLED! 🎵

Welcome to our Sound Reactive fork of WLED. In addition to the features of WLED below, we also support:

- Audio input from several sources including MAX4466, MAX9814, MAX9184, INMP401, INMP441 (for ESP32) and line-in.
- Volume reactive visual effects for ESP32 and ESP8266 devices.
- Frequency reactive visual effects for ESP32 devices.
- UDP sound synchronization with transmit for ESP32 and receive for ESP8266 and ESP32 devices.
- 2D visual effects for ESP32 devices.
- Squelch and gain settings for ESP8266 and ESP32 devices for the volume reactive visual effects.
- 2D settings for ESP32 devices.
- Frequency reactive sliders for ESP32 devices.

We currently have 3 active forks for our Sound Reactive WLED. They are:

- [ESP32 standard version (ESP8266 is no longer supported)](https://github.com/atuline/WLED)
- [ESP32 development version (ESP8266 is no longer supported)](https://github.com/atuline/WLED/tree/dev)
- [ESP8266 version (No longer supported. WILL NOT receive regular updates)](https://github.com/atuline/WLED/tree/ESP8266)

In addition, we have a [Sound Reactive WLED Wiki](https://github.com/atuline/WLED/wiki).

Join Discord to discuss beta testing of our sound reactive fork of WLED:

<a href="https://discord.gg/4CQRmfR"><img src="https://discordapp.com/api/guilds/700041398778331156/widget.png?style=banner2" width="25%"></a>


Join Discord to discuss AirCookie's WLED:

<a href="https://discord.gg/KuqP7NE"><img src="https://discordapp.com/api/guilds/473448917040758787/widget.png?style=banner2" width="25%"></a>



And now onto regular WLED . . .



# Welcome to my project WLED! ✨

A fast and feature-rich implementation of an ESP8266/ESP32 webserver to control NeoPixel (WS2812B, WS2811, SK6812) LEDs or also SPI based chipsets like the WS2801 and APA102!

## ⚙️ Features
- WS2812FX library integrated for over 100 special effects
- FastLED noise effects and 50 palettes
- Modern UI with color, effect and segment controls
- Segments to set different effects and colors to parts of the LEDs
- Settings page - configuration over network
- Access Point and station mode - automatic failsafe AP
- Up to 10 LED outputs per instance
- Support for RGBW strips
- Up to 250 user presets to save and load colors/effects easily, supports cycling through them.
- Presets can be used to automatically execute API calls
- Nightlight function (gradually dims down)
- Full OTA software updatability (HTTP + ArduinoOTA), password protectable
- Configurable analog clock (Cronixie, 7-segment and EleksTube IPS clock support via usermods)
- Configurable Auto Brightness limit for safer operation
- Filesystem-based config for easier backup of presets and settings

## 💡 Supported light control interfaces
- WLED app for [Android](https://play.google.com/store/apps/details?id=com.aircoookie.WLED) and [iOS](https://apps.apple.com/us/app/wled/id1475695033)
- JSON and HTTP request APIs
- MQTT
- Blynk IoT
- E1.31, Art-Net, DDP and TPM2.net
- [diyHue](https://github.com/diyhue/diyHue) (Wled is supported by diyHue, including Hue Sync Entertainment under udp. Thanks to [Gregory Mallios](https://github.com/gmallios))
- [Hyperion](https://github.com/hyperion-project/hyperion.ng)
- UDP realtime
- Alexa voice control (including dimming and color)
- Sync to Philips hue lights
- Adalight (PC ambilight via serial) and TPM2
- Sync color of multiple WLED devices (UDP notifier)
- Infrared remotes (24-key RGB, receiver required)
- Simple timers/schedules (time from NTP, timezones/DST supported)

## 📲 Quick start guide and documentation

See the [documentation on our official site](https://kno.wled.ge)!

[On this page](https://kno.wled.ge/basics/tutorials/) you can find excellent tutorials made by the community and helpful tools to help you get your new lamp up and running!

## 🖼️ User interface
<img src="/images/macbook-pro-space-gray-on-the-wooden-table.jpg" width="50%"><img src="/images/walking-with-iphone-x.jpg" width="50%">

## 💾 Compatible hardware

See [here](https://kno.wled.ge/basics/compatible-hardware)!

## ✌️ Other

Licensed under the MIT license
Credits [here](https://kno.wled.ge/about/contributors/)!

Join the Discord server to discuss everything about WLED!

<a href="https://discord.gg/KuqP7NE"><img src="https://discordapp.com/api/guilds/473448917040758787/widget.png?style=banner2" width="25%"></a>

Check out the WLED [Discourse forum](https://wled.discourse.group)!
You can also send me mails to [dev.aircoookie@gmail.com](mailto:dev.aircoookie@gmail.com), but please only do so if you want to talk to me privately.
If WLED really brightens up your every day, you can [![](https://img.shields.io/badge/send%20me%20a%20small%20gift-paypal-blue.svg?style=flat-square)](https://paypal.me/aircoookie)


*Disclaimer:*
If you are sensitive to photosensitive epilepsy it is not recommended that you use this software.
In case you still want to try, don't use strobe, lighting or noise modes or high effect speed settings.
As per the MIT license, I assume no liability for any damage to you or any other person or equipment.



CHANGELOG.md
## WLED changelog

### Builds after release 0.13.1

#### Build 2203191

-   Fixed sunrise/set calculation (once again)

#### Build 2203190

-   Fixed `/json/cfg` unable to set busses (#2589)
-   Fixed Peek with odd LED counts > 255 (#2586)

#### Build 2203160

-   Version bump to v0.13.2-a0 "Toki"
-   Add ability to skip up to 255 LEDs
-   Dependency version bumps

### WLED release 0.13.1

#### Build 2203150

-   Version bump to v0.13.1 "Toki"
-   Fix persistent preset bug, preventing save of new presets

### WLED release 0.13.0

#### Build 2203142

-   Release of WLED v0.13.0 "Toki"
-   Reduce APA102 hardware SPI frequency to 5Mhz
-   Remove `persistent` parameter in `savePreset()`

### Builds between releases 0.12.0 and 0.13.0

#### Build 2203140

-   Added factory reset by pressing button 0 for >10 seconds
-   Added ability to set presets from DMX Effect mode
-   Simplified label hiding JS in user interface
-   Fixed JSON `{"live":true}` indefinite realtime mode

#### Build 2203080

-   Disabled auto white mode in segments with no RGB bus
-   Fixed hostname string not 0-terminated
-   Fixed Popcorn mode not lighting first LED on pop

#### Build 2203060

-   Dynamic hiding of unused color controls in UI (PR #2567)
-   Removed native Cronixie support and added Cronixie usermod
-   Fixed disabled timed preset expanding calendar
-   Fixed Color Order setting shown for analog busses
-   Fixed incorrect operator (#2566)

#### Build 2203011

-   IR rewrite (PR #2561), supports CCT
-   Added locate button to Time settings
-   CSS fixes and adjustments
-   Consistent Tab indentation in index JS and CSS
-   Added initial contribution style guideline

#### Build 2202222

-   Version bump to 0.13.0-b7 "Toki"
-   Fixed HTTP API commands not applying to all selected segments in some conditions
-   Blynk support is not compiled in by default on ESP32 builds

#### Build 2202210

-   Fixed HTTP API commands not applying to all selected segments if called from JSON
-   Improved Stream effects, no longer rely on LED state and won't fade out at low brightness

#### Build 2202200

-   Added `info.leds.seglc` per-segment light capability info (PR #2552)
-   Fixed `info.leds.rgbw` behavior
-   Segment bounds sync (PR #2547)
-   WebSockets auto reconnection and error handling
-   Disable relay pin by default (PR #2531)
-   Various fixes (ESP32 touch pin 33, floats, PR #2530, #2534, #2538)
-   Deprecated `info.leds.cct`, `info.leds.wv` and `info.leds.rgbw`
-   Deprecated `/url` endpoint

#### Build 2202030

-   Switched to binary format for WebSockets peek (PR #2516)
-   Playlist bugfix
-   Added `extractModeName()` utility function
-   Added serial out (PR #2517)
-   Added configurable baud rate

#### Build 2201260

-   Initial ESP32-C3 and ESP32-S2 support (PRs #2452, #2454, #2502)
-   Full segment sync (PR #2427)
-   Allow overriding of color order by ranges (PR #2463)
-   Added white channel to Peek

#### Build 2112080

-   Version bump to 0.13.0-b6 "Toki"
-   Added "ESP02" (ESP8266 with 2M of flash) to PIO/release binaries

#### Build 2112070

-   Added new effect "Fairy", replacing "Police All"
-   Added new effect "Fairytwinkle", replacing "Two Areas"
-   Static single JSON buffer (performance and stability improvement) (PR #2336)

#### Build 2112030

-   Fixed ESP32 crash on Colortwinkles brightness change
-   Fixed setting picker to black resetting hue and saturation
-   Fixed auto white mode not saved to config

#### Build 2111300

-   Added CCT and white balance correction support (PR #2285)
-   Unified UI slider style
-   Added LED settings config template upload

#### Build 2111220

-   Fixed preset cycle not working from preset called by UI
-   Reintroduced permanent min. and max. cycle bounds

#### Build 2111190

-   Changed default ESP32 LED pin from 16 to 2
-   Renamed "Running 2" to "Chase 2"
-   Renamed "Tri Chase" to "Chase 3"

#### Build 2111170

-   Version bump to 0.13.0-b5 "Toki"
-   Improv Serial support (PR #2334)
-   Button improvements (PR #2284)
-   Added two time zones (PR #2264, 2311)
-   JSON in/decrementing support for brightness and presets
-   Fixed no gamma correction for JSON individual LED control
-   Preset cycle bugfix
-   Removed ledCount
-   LED settings buffer bugfix
-   Network pin conflict bugfix
-   Changed default ESP32 partition layout to 4M, 1M FS

#### Build 2110110

-   Version bump to 0.13.0-b4 "Toki"
-   Added option for bus refresh if off (PR #2259)
-   New auto segment logic
-   Fixed current calculations for virtual or non-linear configs (PR #2262)

#### Build 2110060

-   Added virtual network DDP busses (PR #2245)
-   Allow playlist as end preset in playlist
-   Improved bus start field UX
-   Pin reservations improvements (PR #2214)

#### Build 2109220

-   Version bump to 0.13.0-b3 "Toki"
-   Added segment names (PR #2184)
-   Improved Police and other effects (PR #2184)
-   Reverted PR #1902 (Live color correction - will be implemented as usermod) (PR #2175)
-   Added transitions for segment on/off
-   Improved number of sparks/stars in Fireworks effect with low number of segments
-   Fixed segment name edit pencil disappearing with request
-   Fixed color transition active even if the segment is off
-   Disallowed file upload with OTA lock active
-   Fixed analog invert option missing (PR #2219)

#### Build 2109100

-   Added an auto create segments per bus setting
-   Added 15 new palettes from SR branch (PR #2134)
-   Fixed segment runtime not reset on FX change via HTTP API
-   Changed AsyncTCP dependency to pbolduc fork v1.2.0

#### Build 2108250

-   Added Sync groups (PR #2150)
-   Added JSON API over Serial support
-   Live color correction (PR #1902)

#### Build 2108180

-   Fixed JSON IR remote not working with codes greater than 0xFFFFFF (fixes #2135)
-   Fixed transition 0 edge case

#### Build 2108170

-   Added application level pong websockets reply (#2139)
-   Use AsyncTCP 1.0.3 as it mitigates the flickering issue from 0.13.0-b2

wled00/const.h
#ifndef WLED_CONST_H
#define WLED_CONST_H

/*
 * Readability defines and their associated numerical values + compile-time constants
 */

//Defaults
#define DEFAULT_CLIENT_SSID "Your_Network"
#define DEFAULT_AP_PASS     "zedfy1234"
#define DEFAULT_OTA_PASS    "wledota"

//increase if you need more
#ifndef WLED_MAX_USERMODS
  #ifdef ESP8266
    #define WLED_MAX_USERMODS 4
  #else
    #define WLED_MAX_USERMODS 6
  #endif
#endif

#ifndef WLED_MAX_BUSSES
  #ifdef ESP8266
    #define WLED_MAX_BUSSES 3
  #else
    #ifdef CONFIG_IDF_TARGET_ESP32S2
      #define WLED_MAX_BUSSES 5
    #else
      #define WLED_MAX_BUSSES 10
    #endif
  #endif
#endif

#ifndef WLED_MAX_BUTTONS
  #ifdef ESP8266
    #define WLED_MAX_BUTTONS 2
  #else
    #define WLED_MAX_BUTTONS 4
  #endif
#endif

#ifdef ESP8266
#define WLED_MAX_COLOR_ORDER_MAPPINGS 5
#else
#define WLED_MAX_COLOR_ORDER_MAPPINGS 10
#endif

//Usermod IDs
#define USERMOD_ID_RESERVED               0     //Unused. Might indicate no usermod present
#define USERMOD_ID_UNSPECIFIED            1     //Default value for a general user mod that does not specify a custom ID
#define USERMOD_ID_EXAMPLE                2     //Usermod "usermod_v2_example.h"
#define USERMOD_ID_TEMPERATURE            3     //Usermod "usermod_temperature.h"
#define USERMOD_ID_FIXNETSERVICES         4     //Usermod "usermod_Fix_unreachable_netservices.h"
#define USERMOD_ID_PIRSWITCH              5     //Usermod "usermod_PIR_sensor_switch.h"
#define USERMOD_ID_IMU                    6     //Usermod "usermod_mpu6050_imu.h"
#define USERMOD_ID_FOUR_LINE_DISP         7     //Usermod "usermod_v2_four_line_display.h
#define USERMOD_ID_ROTARY_ENC_UI          8     //Usermod "usermod_v2_rotary_encoder_ui.h"
#define USERMOD_ID_AUTO_SAVE              9     //Usermod "usermod_v2_auto_save.h"
#define USERMOD_ID_DHT                   10     //Usermod "usermod_dht.h"
#define USERMOD_ID_MODE_SORT             11     //Usermod "usermod_v2_mode_sort.h"
#define USERMOD_ID_VL53L0X               12     //Usermod "usermod_vl53l0x_gestures.h"
#define USERMOD_ID_MULTI_RELAY           13     //Usermod "usermod_multi_relay.h"
#define USERMOD_ID_ANIMATED_STAIRCASE    14     //Usermod "Animated_Staircase.h"
#define USERMOD_ID_RTC                   15     //Usermod "usermod_rtc.h"
#define USERMOD_ID_ELEKSTUBE_IPS         16     //Usermod "usermod_elekstube_ips.h"
#define USERMOD_ID_SN_PHOTORESISTOR      17     //Usermod "usermod_sn_photoresistor.h"
#define USERMOD_ID_BATTERY_STATUS_BASIC  18     //Usermod "usermod_v2_battery_status_basic.h"
#define USERMOD_ID_PWM_FAN               19     //Usermod "usermod_PWM_fan.h"
#define USERMOD_ID_BH1750                20     //Usermod "usermod_bh1750.h"
#define USERMOD_ID_SEVEN_SEGMENT_DISPLAY 21     //Usermod "usermod_v2_seven_segment_display.h"
#define USERMOD_RGB_ROTARY_ENCODER       22     //Usermod "rgb-rotary-encoder.h"
#define USERMOD_ID_QUINLED_AN_PENTA      23     //Usermod "quinled-an-penta.h"
#define USERMOD_ID_SSDR                  24     //Usermod "usermod_v2_seven_segment_display_reloaded.h"
#define USERMOD_ID_CRONIXIE              25     //Usermod "usermod_cronixie.h"
#define USERMOD_ID_WIZLIGHTS             26     //Usermod "wizlights.h"
#define USERMOD_ID_WORDCLOCK             27     //Usermod "usermod_v2_word_clock.h"
#define USERMOD_ID_MY9291                28     //Usermod "usermod_MY9291.h"
#define USERMOD_ID_SI7021_MQTT_HA        29     //Usermod "usermod_si7021_mqtt_ha.h"

//Access point behavior
#define AP_BEHAVIOR_BOOT_NO_CONN          0     //Open AP when no connection after boot
#define AP_BEHAVIOR_NO_CONN               1     //Open when no connection (either after boot or if connection is lost)
#define AP_BEHAVIOR_ALWAYS                2     //Always open
#define AP_BEHAVIOR_BUTTON_ONLY           3     //Only when button pressed for 6 sec

//Notifier callMode
#define CALL_MODE_INIT           0     //no updates on init, can be used to disable updates
#define CALL_MODE_DIRECT_CHANGE  1
#define CALL_MODE_BUTTON         2     //default button actions applied to selected segments
#define CALL_MODE_NOTIFICATION   3
#define CALL_MODE_NIGHTLIGHT     4
#define CALL_MODE_NO_NOTIFY      5
#define CALL_MODE_FX_CHANGED     6     //no longer used
#define CALL_MODE_HUE            7
#define CALL_MODE_PRESET_CYCLE   8
#define CALL_MODE_BLYNK          9
#define CALL_MODE_ALEXA         10
#define CALL_MODE_WS_SEND       11     //special call mode, not for notifier, updates websocket only
#define CALL_MODE_BUTTON_PRESET 12     //button/IR JSON preset/macro

//RGB to RGBW conversion mode
#define RGBW_MODE_MANUAL_ONLY     0            //No automatic white channel calculation. Manual white channel slider
#define RGBW_MODE_AUTO_BRIGHTER   1            //New algorithm. Adds as much white as the darkest RGBW channel
#define RGBW_MODE_AUTO_ACCURATE   2            //New algorithm. Adds as much white as the darkest RGBW channel and subtracts this amount from each RGB channel
#define RGBW_MODE_DUAL            3            //Manual slider + auto calculation. Automatically calculates only if manual slider is set to off (0)
#define RGBW_MODE_LEGACY          4            //Old floating algorithm. Too slow for realtime and palette support

//realtime modes
#define REALTIME_MODE_INACTIVE    0
#define REALTIME_MODE_GENERIC     1
#define REALTIME_MODE_UDP         2
#define REALTIME_MODE_HYPERION    3
#define REALTIME_MODE_E131        4
#define REALTIME_MODE_ADALIGHT    5
#define REALTIME_MODE_ARTNET      6
#define REALTIME_MODE_TPM2NET     7
#define REALTIME_MODE_DDP         8

//realtime override modes
#define REALTIME_OVERRIDE_NONE    0
#define REALTIME_OVERRIDE_ONCE    1
#define REALTIME_OVERRIDE_ALWAYS  2

//E1.31 DMX modes
#define DMX_MODE_DISABLED         0            //not used
#define DMX_MODE_SINGLE_RGB       1            //all LEDs same RGB color (3 channels)
#define DMX_MODE_SINGLE_DRGB      2            //all LEDs same RGB color and master dimmer (4 channels)
#define DMX_MODE_EFFECT           3            //trigger standalone effects of WLED (11 channels)
#define DMX_MODE_MULTIPLE_RGB     4            //every LED is addressed with its own RGB (ledCount * 3 channels)
#define DMX_MODE_MULTIPLE_DRGB    5            //every LED is addressed with its own RGB and share a master dimmer (ledCount * 3 + 1 channels)
#define DMX_MODE_MULTIPLE_RGBW    6            //every LED is addressed with its own RGBW (ledCount * 4 channels)

//Light capability byte (unused) 0bRCCCTTTT
//bits 0/1/2/3: specifies a type of LED driver. A single "driver" may have different chip models but must have the same protocol/behavior
//bits 4/5/6: specifies the class of LED driver - 0b000 (dec. 0-15)  unconfigured/reserved
//                                              - 0b001 (dec. 16-31) digital (data pin only)
//                                              - 0b010 (dec. 32-47) analog (PWM)
//                                              - 0b011 (dec. 48-63) digital (data + clock / SPI)
//                                              - 0b100 (dec. 64-79) unused/reserved
//                                              - 0b101 (dec. 80-95) virtual network busses
//                                              - 0b110 (dec. 96-111) unused/reserved
//                                              - 0b111 (dec. 112-127) unused/reserved
//bit 7 is reserved and set to 0

#define TYPE_NONE                 0            //light is not configured
#define TYPE_RESERVED             1            //unused. Might indicate a "virtual" light
//Digital types (data pin only) (16-31)
#define TYPE_WS2812_1CH          20            //white-only chips
#define TYPE_WS2812_WWA          21            //amber + warm + cold white
#define TYPE_WS2812_RGB          22
#define TYPE_GS8608              23            //same driver as WS2812, but will require signal 2x per second (else displays test pattern)
#define TYPE_WS2811_400KHZ       24            //half-speed WS2812 protocol, used by very old WS2811 units
#define TYPE_SK6812_RGBW         30
#define TYPE_TM1814              31
//"Analog" types (PWM) (32-47)
#define TYPE_ONOFF               40            //binary output (relays etc.)
#define TYPE_ANALOG_1CH          41            //single channel PWM. Uses value of brightest RGBW channel
#define TYPE_ANALOG_2CH          42            //analog WW + CW
#define TYPE_ANALOG_3CH          43            //analog RGB
#define TYPE_ANALOG_4CH          44            //analog RGBW
#define TYPE_ANALOG_5CH          45            //analog RGB + WW + CW
//Digital types (data + clock / SPI) (48-63)
#define TYPE_WS2801              50
#define TYPE_APA102              51
#define TYPE_LPD8806             52
#define TYPE_P9813               53
//Network types (master broadcast) (80-95)
#define TYPE_NET_DDP_RGB         80            //network DDP RGB bus (master broadcast bus)
#define TYPE_NET_E131_RGB        81            //network E131 RGB bus (master broadcast bus)
#define TYPE_NET_ARTNET_RGB      82            //network ArtNet RGB bus (master broadcast bus)

#define IS_DIGITAL(t) ((t) & 0x10) //digital are 16-31 and 48-63
#define IS_PWM(t)     ((t) > 40 && (t) < 46)
#define NUM_PWM_PINS(t) ((t) - 40) //for analog PWM 41-45 only
#define IS_2PIN(t)      ((t) > 47)

//Color orders
#define COL_ORDER_GRB             0           //GRB(w),defaut
#define COL_ORDER_RGB             1           //common for WS2811
#define COL_ORDER_BRG             2
#define COL_ORDER_RBG             3
#define COL_ORDER_BGR             4
#define COL_ORDER_GBR             5
#define COL_ORDER_MAX             5


//Button type
#define BTN_TYPE_NONE             0
#define BTN_TYPE_RESERVED         1
#define BTN_TYPE_PUSH             2
#define BTN_TYPE_PUSH_ACT_HIGH    3
#define BTN_TYPE_SWITCH           4
#define BTN_TYPE_PIR_SENSOR       5
#define BTN_TYPE_TOUCH            6
#define BTN_TYPE_ANALOG           7
#define BTN_TYPE_ANALOG_INVERTED  8

//Ethernet board types
#define WLED_NUM_ETH_TYPES        8

#define WLED_ETH_NONE             0
#define WLED_ETH_WT32_ETH01       1
#define WLED_ETH_ESP32_POE        2
#define WLED_ETH_WESP32           3
#define WLED_ETH_QUINLED          4
#define WLED_ETH_TWILIGHTLORD     5
#define WLED_ETH_ESP32DEUX        6

//Hue error codes
#define HUE_ERROR_INACTIVE        0
#define HUE_ERROR_UNAUTHORIZED    1
#define HUE_ERROR_LIGHTID         3
#define HUE_ERROR_PUSHLINK      101
#define HUE_ERROR_JSON_PARSING  250
#define HUE_ERROR_TIMEOUT       251
#define HUE_ERROR_ACTIVE        255

//Segment option byte bits
#define SEG_OPTION_SELECTED       0
#define SEG_OPTION_REVERSED       1
#define SEG_OPTION_ON             2
#define SEG_OPTION_MIRROR         3            //Indicates that the effect will be mirrored within the segment
#define SEG_OPTION_ROTATED2D      4            //Indicates that the effect will be rotated 90d within the segment
// #define SEG_OPTION_NONUNITY       4            //Indicates that the effect does not use FRAMETIME or needs getPixelColor
#define SEG_OPTION_FREEZE         5            //Segment contents will not be refreshed
#define SEG_OPTION_REVERSED2D     6            //ewowi20210705
#define SEG_OPTION_TRANSITIONAL   7

//Segment differs return byte
#define SEG_DIFFERS_BRI        0x01
#define SEG_DIFFERS_OPT        0x02
#define SEG_DIFFERS_COL        0x04
#define SEG_DIFFERS_FX         0x08
#define SEG_DIFFERS_BOUNDS     0x10
#define SEG_DIFFERS_GSO        0x20
#define SEG_DIFFERS_SEL        0x80

//Playlist option byte
#define PL_OPTION_SHUFFLE      0x01

// WLED Error modes
#define ERR_NONE         0  // All good :)
#define ERR_EEP_COMMIT   2  // Could not commit to EEPROM (wrong flash layout?)
#define ERR_JSON         9  // JSON parsing failed (input too large?)
#define ERR_FS_BEGIN    10  // Could not init filesystem (no partition?)
#define ERR_FS_QUOTA    11  // The FS is full or the maximum file size is reached
#define ERR_FS_PLOAD    12  // It was attempted to load a preset that does not exist
#define ERR_FS_IRLOAD   13  // It was attempted to load an IR JSON cmd, but the "ir.json" file does not exist
#define ERR_FS_GENERAL  19  // A general unspecified filesystem error occured
#define ERR_OVERTEMP    30  // An attached temperature sensor has measured above threshold temperature (not implemented)
#define ERR_OVERCURRENT 31  // An attached current sensor has measured a current above the threshold (not implemented)
#define ERR_UNDERVOLT   32  // An attached voltmeter has measured a voltage below the threshold (not implemented)

//Timer mode types
#define NL_MODE_SET               0            //After nightlight time elapsed, set to target brightness
#define NL_MODE_FADE              1            //Fade to target brightness gradually
#define NL_MODE_COLORFADE         2            //Fade to target brightness and secondary color gradually
#define NL_MODE_SUN               3            //Sunrise/sunset. Target brightness is set immediately, then Sunrise effect is started. Max 60 min.

//EEPROM size
#ifdef ESP8266
#define EEPSIZE 3300  //Maximum is 4096
#else // ESP8266
#define EEPSIZE 3300  //Maximum is 4096
#endif // ESP32


#define NTP_PACKET_SIZE 48

//maximum number of rendered LEDs - this does not have to match max. physical LEDs, e.g. if there are virtual busses
#ifndef MAX_LEDS
#ifdef ESP8266
#define MAX_LEDS 1664 //can't rely on memory limit to limit this to 1600 LEDs
#else
#define MAX_LEDS 8192
#endif
#endif

#ifndef MAX_LED_MEMORY
#ifdef ESP8266
#define MAX_LED_MEMORY 4000
#else
#define MAX_LED_MEMORY 64000
#endif
#endif

#ifndef MAX_LEDS_PER_BUS
#define MAX_LEDS_PER_BUS 4096
#endif

// string temp buffer (now stored in stack locally)
#ifdef ESP8266
#define SETTINGS_STACK_BUF_SIZE 2048
#else
#define SETTINGS_STACK_BUF_SIZE 3096
#endif

#ifdef WLED_USE_ETHERNET
  #define E131_MAX_UNIVERSE_COUNT 20
#else
  #ifdef ESP8266
    #define E131_MAX_UNIVERSE_COUNT 9
  #else
    #define E131_MAX_UNIVERSE_COUNT 12
  #endif
#endif

#ifndef ABL_MILLIAMPS_DEFAULT
  #define ABL_MILLIAMPS_DEFAULT 850  // auto lower brightness to stay close to milliampere limit
#else
  #if ABL_MILLIAMPS_DEFAULT < 250  // make sure value is at least 250
   #define ABL_MILLIAMPS_DEFAULT 250
  #endif
#endif

// PWM settings
#ifndef WLED_PWM_FREQ
#ifdef ESP8266
  #define WLED_PWM_FREQ    880 //PWM frequency proven as good for LEDs
#else
  #define WLED_PWM_FREQ  19531
#endif
#endif

#define TOUCH_THRESHOLD 32 // limit to recognize a touch, higher value means more sensitive

// Size of buffer for API JSON object (increase for more segments)
#ifdef ESP8266
  #define JSON_BUFFER_SIZE 10240
#else
  #define JSON_BUFFER_SIZE 20480
#endif

#ifdef WLED_USE_DYNAMIC_JSON
  #define MIN_HEAP_SIZE JSON_BUFFER_SIZE+512
#else
  #define MIN_HEAP_SIZE 4096
#endif

// Maximum size of node map (list of other WLED instances)
#ifdef ESP8266
  #define WLED_MAX_NODES 24
#else
  #define WLED_MAX_NODES 150
#endif

//this is merely a default now and can be changed at runtime
#ifndef LEDPIN
#ifdef ESP8266
  #define LEDPIN 2    // GPIO2 (D4) on Wemod D1 mini compatible boards
#else
  #define LEDPIN 16   // Changed from 16 to restore compatibility with ESP32-pico
#endif
#endif

#ifdef WLED_ENABLE_DMX
#if (LEDPIN == 2)
  #undef LEDPIN
  #define LEDPIN 1
  #warning "Pin conflict compiling with DMX and LEDs on pin 2. The default LED pin has been changed to pin 1."
#endif
#endif

#ifndef DEFAULT_LED_COUNT
  #define DEFAULT_LED_COUNT 30
#endif

#define INTERFACE_UPDATE_COOLDOWN 2000 //time in ms to wait between websockets, alexa, and MQTT updates

#endif
